examples/crowdfunding/test/failing-tests.js

```diff
@@ -0,0 +1,437 @@
+      payFlags: {}
+    });
+
+    appArgs = [stringToBytes('donate')];
+    donateTxGroup = [
+      {
+        type: types.TransactionType.CallNoOpSSC,
+        sign: types.SignType.SecretKey,
+        fromAccount: donor.account,
+        appId: applicationId,
+        payFlags: { totalFee: 1000 },
+        appArgs: appArgs
+      },
+      {
+        type: types.TransactionType.TransferAlgo,
+        sign: types.SignType.SecretKey,
+        fromAccount: donor.account,
+        toAccountAddr: escrow.address,
+        amountMicroAlgos: goal - 1e6,
+        payFlags: { totalFee: 1000 }
+      }
+    ];
+  });
+
+  function updateAndOptIn () {
+    // update application with correct escrow account address
+    appArgs = [addressToPk(escrowAddress)]; // converts algorand address to Uint8Array
+
+    runtime.updateApp(
+      creator.address,
+      applicationId,
+      approvalProgram,
+      clearProgram,
+      {}, { appArgs: appArgs });
+
+    // opt-in to app
+    runtime.optInToApp(creator.address, applicationId, {}, {});
+    runtime.optInToApp(donor.address, applicationId, {}, {});
+  }
+
+  this.afterEach(async function () {
+    creator = new StoreAccount(initialCreatorBalance);
+    donor = new StoreAccount(initialDonorBalance);
+    runtime = new Runtime([master, creator, escrow, donor]);
+  });
+
+  it('should fail donation if donor has insufficient balance', () => {
+    updateAndOptIn();
+    appArgs = [stringToBytes('donate')];
+    const donationAmount = initialDonorBalance + 1000;
+    // Atomic Transaction (Stateful Smart Contract call + Payment Transaction)
+    const txGroup = [
+      {
+        type: types.TransactionType.CallNoOpSSC,
+        sign: types.SignType.SecretKey,
+        fromAccount: donor.account,
+        appId: applicationId,
+        payFlags: { totalFee: 1000 },
+        appArgs: appArgs
+      },
+      {
+        type: types.TransactionType.TransferAlgo,
+        sign: types.SignType.SecretKey,
+        fromAccount: donor.account,
+        toAccountAddr: escrow.address,
+        amountMicroAlgos: donationAmount,
+        payFlags: { totalFee: 1000 }
+      }
+    ];
+
+    assert.throws(() => runtime.executeTx(txGroup), 'RUNTIME_ERR1401: Cannot withdraw');
+  });
+
+  it('should fail donation if timestamp is after endDate', () => {
+    updateAndOptIn();
+    // set timestamp to after of endDate
+    runtime.setRoundAndTimestamp(5, endDate.getSeconds() + 100);
+
+    assert.throws(() => runtime.executeTx(donateTxGroup), rejectMsg);
+  });
+
+  it('should fail donation if timestamp is before of beginDate', () => {
+    updateAndOptIn();
+    // set timestamp to out of endDate
+    runtime.setRoundAndTimestamp(5, beginDate.getSeconds() - 100);
+
+    assert.throws(() => runtime.executeTx(donateTxGroup), rejectMsg);
+  });
+
+  it('should fail if goal is met, and donor tries to reclaim funds', () => {
+    updateAndOptIn();
+    runtime.executeTx(donateTxGroup);
+    runtime.executeTx(donateTxGroup);
```

Comment:
Why are we doing this twice here? Can you please add a comment?

Reply:
Also, i wanted to understand, why can't we just fund the escrow on our own in this case while preparing the state (with a simple `algotransfer` txn, and setting "Total" field in smart contract). Is it necessary to go through the `donate` tx? Currently in this test there is dependency on `donate:` branch, is it possible to avoid that?

Reply:
we are doing this to meet the goal of the donation.

we cannot do that because contract has variables which needs to be updated. we are using this donate: branch because we cannot test claim if we haven't donated to escrow.
changing this will be like updating app without even creating it. dependency will be there to properly test claim part.

Reply:
I was thinking that we can prepare the state without calling the smart contract. eg.
```
runtime.transferAlgo(from, to, ..) // amount = goal, to = escrow
runtime.setGlobalState(appId, "Total", `${goal}`)
creator.setLocalState(appId, "MyAmountGiven",`${goal}`)
```
This way we prepare the correct state before doing claim/reclaim. I just think that `runtime.executeTx(donateTxGroup);` can fail due to number of reasons (eg - mistake in `donate:` in written smart contract, sender not creator, txGroup incorrect, common property incorrect ..etc). So if we can avoid that, and just prepare the correct state, it would be good (as we will be able to test claim/reclaim now with less failures due to dependency). 
Anyways, just a thought. This works too.